lucille-core/src/main/java/com/kmwllc/lucille/stage/ApplyJSoup.java

```diff
@@ -0,0 +1,137 @@
+ * to UTF-8 immediately. Has no effect if stringField is provided.
+ * @see <a href="https://docs.oracle.com/javase/8/docs/api/java/nio/charset/Charset.html"> Charsets </a> for information on supported charsets and conventions for 
+ * specifying them.
+ * </p>
+ * <b>destinationFields</b> (Map<String, String>) : defines a mapping from destination fields to css selectors.
+ * If a destination field already exists in the processed document it is overwritten. Otherwise, they are created. If a selector returns 
+ * multiple elements the destination field receives a list of processed elements. If a selector returns no elements for a document then the 
+ * destination field is not created
+ * @see <a href="https://jsoup.org/cookbook/extracting-data/selector-syntax"> CSS selectors </a> for information on supported selectors
+ * </p>
+ */
+public class ApplyJSoup extends Stage {
+
+  private final String filePathField;
+  private final Map<String, Object> destinationFields;
+  private final String byteArrayField;
+  private final String stringField;
+  private final String charset;
+
+  public ApplyJSoup(Config config) {
+    super(config, new StageSpec().withOptionalProperties("filePathField", "byteArrayField", "stringField", "charset")
+        .withRequiredParents("destinationFields"));
+
+    this.destinationFields = config.getConfig("destinationFields").root().unwrapped();
+    this.filePathField = ConfigUtils.getOrDefault(config, "filePathField", null);
+    this.byteArrayField = ConfigUtils.getOrDefault(config, "byteArrayField", null);
+    this.charset = ConfigUtils.getOrDefault(config, "charset", null);
+    this.stringField = ConfigUtils.getOrDefault(config, "stringField", null);
+  }
+
+  @Override
+  public void start() throws StageException {
+    if (!((this.filePathField != null ^ this.byteArrayField != null ^ this.stringField != null
+        ^ (this.filePathField != null && this.byteArrayField != null && this.stringField != null)))) {
+      throw new StageException("Stage must have one and only one of filePathField, stringField, or byteArrayField specified");
+    }
+
+    for (Map.Entry<String, Object> entry : destinationFields.entrySet()) {
+      Object value = entry.getValue();
+      if (value instanceof Map) {
+        Map<String, String> map = (Map<String, String>) value;
+        if (map.keySet().size() != 2) {
+          throw new StageException("map must contain two fields");
+        }
+        if (!map.containsKey("attribute") || !map.containsKey("selector")) {
+          throw new StageException("map must contain a attribute and selector fields");
+        }
+      } else if (!(value instanceof String)) {
+        throw new StageException("destinationFields mapping can only have strings or a map as its values");
+      }
+    }
+  }
+
+  private String activeField() {
```

Comment:
We can determine and store the active field in the constructor of this stage -- I don't think we need to determine it dynamically on each call to processDocument